File: backend/app/schemas/opportunity.py
"""
Pydantic schemas for Opportunity — request validation and response serialization.

IDs are str (UUID stored as String(36) in the model) for SQLite compatibility.
"""
from pydantic import BaseModel, field_validator
from typing import Optional, List
from datetime import date, datetime


VALID_CATEGORIES = {
    "scholarship", "fellowship", "accelerator", "vc_program", "grant",
    "competition", "conference", "exhibition", "exchange_program",
    "travel_program", "government_scheme", "giveaway", "other"
}

VALID_STATUSES = {
    "saved", "planning", "applied", "interview", "accepted", "rejected", "waitlisted"
}


class OpportunityBase(BaseModel):
    title: str
    organization: Optional[str] = None
    description: Optional[str] = None
    category: Optional[str] = None
    tags: Optional[List[str]] = []
    country: Optional[List[str]] = []
    is_remote: Optional[bool] = False
    women_friendly: Optional[bool] = False
    india_eligible: Optional[bool] = False
    student_eligible: Optional[bool] = False
    age_limit: Optional[str] = None
    eligibility: Optional[str] = None
    funding_amount: Optional[str] = None
    application_fee: Optional[str] = None
    link: str
    source_url: Optional[str] = None
    source_name: Optional[str] = None
    deadline: Optional[date] = None
    is_expired: Optional[bool] = False


class OpportunityCreate(OpportunityBase):
    pass


class OpportunityUpdate(BaseModel):
    title: Optional[str] = None
    organization: Optional[str] = None
    description: Optional[str] = None
    category: Optional[str] = None
    tags: Optional[List[str]] = None
    country: Optional[List[str]] = None
    is_remote: Optional[bool] = None
    women_friendly: Optional[bool] = None
    india_eligible: Optional[bool] = None
    student_eligible: Optional[bool] = None
    age_limit: Optional[str] = None
    eligibility: Optional[str] = None
    funding_amount: Optional[str] = None
    application_fee: Optional[str] = None
    deadline: Optional[date] = None
    is_expired: Optional[bool] = None


class OpportunityRead(OpportunityBase):
    id: str
    created_at: datetime
    updated_at: datetime

    @field_validator("id", mode="before")
    @classmethod
    def coerce_id(cls, v):
        return str(v)

    @field_validator("tags", "country", mode="before")
    @classmethod
    def ensure_list(cls, v):
        if v is None:
            return []
        if isinstance(v, str):
            import json
            try:
                return json.loads(v)
            except Exception:
                return []
        return v

    model_config = {"from_attributes": True}


class OpportunityListResponse(BaseModel):
    total: int
    page: int
    page_size: int
    results: List[OpportunityRead]

File: frontend/src/types/opportunity.ts
export type OpportunityCategory =
    | "scholarship"
    | "fellowship"
    | "accelerator"
    | "vc_program"
    | "grant"
    | "competition"
    | "conference"
    | "exhibition"
    | "exchange_program"
    | "travel_program"
    | "government_scheme"
    | "giveaway"
    | "other";

export type ApplicationStatus =
    | "saved"
    | "planning"
    | "applied"
    | "interview"
    | "accepted"
    | "rejected"
    | "waitlisted";

export interface Opportunity {
    id: string;
    title: string;
    organization: string | null;
    description: string | null;
    category: OpportunityCategory;
    tags: string[];
    country: string[];
    is_remote: boolean;
    women_friendly: boolean;
    india_eligible: boolean;
    student_eligible: boolean;
    age_limit: string | null;
    eligibility: string | null;
    funding_amount: string | null;
    application_fee: string | null;
    link: string;
    source_name: string | null;
    deadline: string | null; // ISO date string
    is_expired: boolean;
    created_at: string;
    updated_at: string;
}

export interface OpportunityListResponse {
    total: number;
    page: number;
    page_size: number;
    results: Opportunity[];
}

export interface Application {
    id: string;
    opportunity_id: string;
    status: ApplicationStatus;
    notes: string | null;
    priority: number;
    applied_at: string | null;
    reminder_date: string | null;
    document_links: string | null;
    created_at: string;
    updated_at: string;
    opportunity: Opportunity | null;
}

export interface OpportunityFilters {
    search?: string;
    category?: string;
    country?: string;
    tag?: string;
    women_friendly?: boolean;
    india_eligible?: boolean;
    student_eligible?: boolean;
    is_remote?: boolean;
    is_expired?: boolean;
    page?: number;
    page_size?: number;
}
